bin/main.rb

```diff
@@ -1,70 +1,227 @@
 #!/usr/bin/env ruby
-p 'Welcome! This is Tic Tac Toe Game. Have Fun!'
+require './lib/player.rb'
+require './lib/canvas.rb'
 
-p 'Please enter player names'
-player_one = gets.chomp
-puts "#{player_one} you'll play as X"
+# validate and print player choice
+class UserInterface
+  def user_info
+    canvas = Canvas.new
+    @arr = canvas.array_canvas
 
-puts ' '
-p 'Player 2, please enter your name too'
-player_two = gets.chomp
-puts "#{player_two} you'll play as O"
+    puts 'Tic Tac Toe Game!'
 
-puts ' '
-puts 'Game Board'
-
-canvas = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
-canvas.each do |i|
-  print i
-  puts
-end
-puts ' '
-
-start_game = true
-slots = 0
-
-player_one_choices = []
-player_two_choices = []
-
-while start_game
-  puts "#{player_one}, choose from any available slot(s), from 1 to 9"
-  choice_one = gets.chomp
-  state = true
-  while state
-    if choice_one.to_i.positive? && choice_one.to_i < 10
-      player_one_choices.push(choice_one)
-      puts "Mark number-slot chosen by #{player_one} as X on the canvas"
-      puts ' '
-
-      state = false
-      slots += 1
-      next
-    else
-      puts 'Please enter a valid value between 1 and 9'
-      break
+    puts
+    puts 'Game Board'
+    @arr.each do |i|
+      print i
+      puts
     end
+
+    @player_one = Player.new
+    @player_two = Player.new
+
+    print 'Enter name for Player One: '
+    @player_one.name = gets.chomp
+    @player_one.name = @player_one.name.capitalize
+
+    print 'Enter name for Player Two: '
+    @player_two.name = gets.chomp
+    @player_two.name = @player_two.name.capitalize
```

Comment:
#### Implementation details

- [ ] The code should respect the single responsibility rule for methods (i.e. each method does exactly one thing).

Your **user_info** method is not only storing the player's info is also requesting the player for info, let's create separate methods for each responsibility;

- [ ] Keep your code DRY (Don't repeat yourself)

 lines 23 to 25 and 27 to 29 are repetitive I think you could save some lines by creating a method for it and call it 2 times within your main.rb file